### PR TITLE
Feature/MariaDB dialect

### LINKED ISSUE #5840: MariaDB into existing MySQL, or new dialect for CREATE/DROP [OR REPLACE] ?
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

MariaDB has some small syntactic differences compared to MySQL, but for the most part should be very similar. I have run into issues where the linter throws `unparsable section` errors because of the presence of `CREATE OR REPLACE TABLE` statements, because MySQL doesn't support the OR REPLACE syntax. 

There are many more instances where parsing errors happen due to small differences between these dialects. 

Would it be preferrable to add to the existing MySQL syntax, or to create a whole new dialect specifically for the small MariaDB changes?i

### Use case

`mariadb` dialect added that borrows heavily from `mysql` but with syntax differences accounted for.

### Dialect

MySQL

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Xemptuous**
### Brief summary of the change made

fixes #5840 by adding new MariaDB dialect based off mysql.

This PR does a few things:

- Adds core functionality for MariaDB
- Adds partitions/subpartitions and partition_option/subpartition_option to table DDL for MySQL
- Adds `REGEXP` into `LikeGrammar` for MySQL

### Are there any other side effects of this change that we should be aware of?

Potentially more complex tests would be required to ensure the added sections for partitions do not affect other code in larger scripts.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by WittierDinosaur** [comment]
is this meant to be in draft?

**Comment by Xemptuous** [comment]
> is this meant to be in draft?

Yea I set it as a draft so I can go through other areas that MariaDB's dialect potentially differs from MySQL in a critical way, while still providing visibility.

**Comment by WittierDinosaur** [comment]
The PR is already pretty massive - feel free to add to this incrementally later. The first drop doesn't need to be a full implementation

File: test/fixtures/dialects/mariadb/.sqlfluff
Comment (alanmcruickshank) [inline test/fixtures/dialects/mariadb/.sqlfluff:2]: Shouldn't this be mariadb?

Comment (Xemptuous) [reply test/fixtures/dialects/mariadb/.sqlfluff:2]: Yes, I just copied all the mysql tests over to mariadb and kept them as is. Will change this as well

File: src/sqlfluff/dialects/dialect_mysql.py
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_mysql.py:384]: I don't love this aliasing here. There's nothing inherently wrong with the pattern. But we don't use it elsewhere, so I think it overall makes it less readable.

Comment (Xemptuous) [reply src/sqlfluff/dialects/dialect_mysql.py:384]: I'm not a big fan either; it was the only way I was able to get the linter to accept the code due to column lengths without it.

Comment (Xemptuous) [reply src/sqlfluff/dialects/dialect_mysql.py:384]: I agree, but in any of my limited attempts to simply break the sequences out into their own segments, I end up getting test failures. I'm pretty limited on time right now, but I'm not opposed to getting rid of the partition/subpartition options. It was just a lacking feature I thought I'd implement while I was at it, but I don't really use it personally, so I don't mind stripping it to get this PR pushed through. Lmk your thoughts

File: src/sqlfluff/dialects/dialect_mariadb_keywords.py
Comment (pprkut) [inline src/sqlfluff/dialects/dialect_mariadb_keywords.py:4]: I added slightly more documentation to my list of reserved keywords, especially where to get the unreserved keywords from. Feel free to steal it :)

https://github.com/pprkut/sqlfluff/blob/mariadb-dialect/src/sqlfluff/dialects/dialect_mariadb_keywords.py#L1

File: src/sqlfluff/dialects/dialect_mariadb_keywords.py
Comment (pprkut) [inline src/sqlfluff/dialects/dialect_mariadb_keywords.py:706]: The `mysql` dialect has a couple extra "keywords" at the end: https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_mysql_keywords.py#L784

Are you sure those aren't needed, given that this extends the mysql dialect?

**Review by pprkut — COMMENTED** [review]
I don't know what the policy is for test fixtures, but is there a need to copy everything from the mysql dialect? IMHO, just adding fixtures for the features that differ from the mysql dialect should be enough, no?

**Review by WittierDinosaur — APPROVED** [review]
This is functional, and working. LGTM. I don't love the shorthand in the mysql dialect, but that's cosmetic, so I won't hold this up. We can always break this down more in the future.

**Comment by WittierDinosaur** [comment]
@Xemptuous ready for this to go, but you've got some yml that needs regenerating after a recent change that was merged ahead of yours

File: src/sqlfluff/dialects/dialect_mariadb_keywords.py
Comment (Xemptuous) [reply src/sqlfluff/dialects/dialect_mariadb_keywords.py:4]: Thanks! Adding in upcoming commit :)

File: src/sqlfluff/dialects/dialect_mariadb_keywords.py
Comment (Xemptuous) [reply src/sqlfluff/dialects/dialect_mariadb_keywords.py:706]: They weren't listed in the mariadb repo or docs as keywords, so I don't know if they're supported or not. Might be worth exploring, as I've never seen or used some of these (e.g., `Zone` or `XID`)

**Pull request merged**
